### ISSUE TITLE
Adding request script in McM and save the log

### ISSUE THREAD

**Issue by pgunnell** [cms-PdmV/cmsPdmV#963]
Middle priority: will help to speed up approval procedure from GEN

- Request script will be provided by GEN
- Add a button in McM for MC contact to run the script (independent on validation)
- The output of the request script should be saved somewhere (output can vary between 5 lines up to 20-30) and accessible to MC contacts and GEN conveners
- If the button is clicked many times, only the last log is stored, no need for history

**Comment by srimanob** [comment]
I'm not sure I agree on the point to have additional button to click / step to add to GEN contact. What I would like to propose is
+ This is part of validation, run after CMSSW area is created
+ After script is running, you need to have a final line to say it is "OK". Validation script then check the log, and exist from validation if the final conclusion is not OK.
+ As it runs on condor, it should be able to access eos, then the log can put to some specific place. No need to implement anything on McM machine side
+ Recommendation to GEN contact should keep the same, they should try to run locally first before putting to validation. If not, and it fails in validation, they waste their time (on waiting for validation report).

No need of extra step / No extra button / Log is kept on EOS

**Comment by efeyazgan** [comment]
Hi Phat, 

Thanks for the input. See inline:

> I'm not sure I agree on the point to have additional button to click / step to add to GEN contact. What I would like to propose is
> 
>     * This is part of validation, run after CMSSW area is created
> 
The contact and/or the gen conveners and/or gen operators should have the option to run it at any stage. Before validation will be the most useful one in general. If we mix it with validation for a 2 minute check we may need to wait for up to several days depending on the process. And in that case, it would be better not to have it in mcm then. 

>     * After script is running, you need to have a final line to say it is "OK". 
This is a good suggestion - we can do it for many cases but not all (at least not all immediately). 

> Validation script then check the log, and exist from validation if the final conclusion is not OK.
> 

This is a fast process (apart from the slha ones) and if we mix it with validation contacts will lose unnecessary time as indicated above. 

>     * As it runs on condor, it should be able to access eos, then the log can put to some specific place. No need to implement anything on McM machine side
> 

No, it should not run on condor (It would be like submitting a job to list a directory with "ls"). There is nothing to run in parallel or slow. (in future we can try to parallelize the slha ones but that might be overkill).  This is a simple fast script.

>     * Recommendation to GEN contact should keep the same, they should try to run locally first before putting to validation. If not, and it fails in validation, they waste their time (on waiting for validation report).

Yes, so before running validation in mcm they should run the request checking script. 

> 
> 
> No need of extra step / No extra button / Log is kept on EOS

I don't agree. We need an extra step an extra button and log to be linked to each request in MCM. If you want we can sit down together and I can run the script live and you will better understand when you have time. In any case, if you do not want it in mcm please let me know soon, then we will do it in a different webpage. Sorry, I do not see any other solution than to have it run in a webpage with a button producing a link to a log.

**Comment by srimanob** [comment]
I think we don't agree on the point that this should be part of validation. I don't want to have extra step in McM for GEN contact. It's too complicated even now (from the survey)

As I said, if GEN contact want to waste their time, no local test should run. The test in McM should be just the final x-check. I don't see the point that if it passes the local, why it will fail in condor.

I ran the script before I type the comment;
(132) genproductions/bin/utils: python request_fragment_check.py --prepid HIG-RunIIFall17wmLHEGS-03157 --bypass_status
---> 1 requests will be checked:
 
if that fails, it because a release had been setup first ...
Error: Cannot authenticate to: https://cms-pdmv-dev.cern.ch/mcm/

Error: Cannot authenticate to: https://cms-pdmv-int.cern.ch/mcm/

Error: Cannot authenticate to: https://cms-pdmv-dev.cern.ch/mcm/

Found a cookie file at /afs/cern.ch/user/s/srimanob/private/prod-cookie.txt. Make sure it's not expired!
Using sso-cookie file /afs/cern.ch/user/s/srimanob/private/prod-cookie.txt
HIG-RunIIFall17wmLHEGS-03157

***********************************************************************************
HIG-RunIIFall17wmLHEGS-03157    Status= done
GluGluToHHTo2B2WToLNu2J_node_SM_TuneCP5_13TeV-madgraph_correctedcfg
/cvmfs/cms.cern.ch/phys_generator/gridpacks/2017/13TeV/madgraph/V5_2.4.2/GF_HH_SM/v1/GF_HH_SM_slc6_amd64_gcc481_CMSSW_7_1_30_tarball.tar.xz
/eos/cms/store/group/phys_generator/cvmfs/gridpacks/2017/13TeV/madgraph/V5_2.4.2/GF_HH_SM/v1/GF_HH_SM_slc6_amd64_gcc481_CMSSW_7_1_30_tarball.tar.xz
generate p p > h h

set group_subprocesses Auto
set ignore_six_quark_processes False

 [IMPORTANT WARNING] bwcutoff set to 15000.0. Note that large bwcutoff values can cause problems in production.
* [WARNING] At least one of the MG5_aMC@NLO tmpdir patches is missing.
*           --> Please check that you use:
*               >=  CMSSW_7_1_32_patch1 in 7_1_X or
*               >= CMSSW_9_3_9_patch1 in 9_3_X or
*               >= 10_1_3 in 10_1_X or
*               >= CMSSW_10_2_0_pre2 in 10_2_X.
*           Your request uses CMSSW_9_3_9_patch1 :
*           If you are not using a proper CMSSW version, please switch to that or
*           re-create the gridpack using the updated genproductions area
*
-------------------------MG5_aMC LO/MLM Many Threads Patch Check --------------------------------------
* [OK] MG5_aMC@NLO LO nthreads patch OK in EOS
* [OK] MG5_aMC@NLO LO nthreads patch OK in CVMFS
-------------------------EOF MG5_aMC LO/MLM Many Threads Patch Check ----------------------------------
*
* [OK] no known inconsistency in the fragment w.r.t. the name of the dataset madgraph
* [OK] Tune configuration probably OK in the fragment
* [WARNING] Filters in the fragment but filter efficiency = 1
***********************************************************************************

**Comment by srimanob** [comment]
For the log, I think in the meeting we are talking about EOS. It should not be an issue to link to McM (but it's just a link to EOS file, and maybe can't show). It you need it to be shown in McM, something difference must be done.

**Comment by srimanob** [comment]
"If we mix it with validation for a 2 minute check we may need to wait for up to several days depending on the process. And in that case, it would be better not to have it in mcm then."
is not what I propose. 

If this part of validation fail, the validation job should quit from condor immediately. No need to download the genfragment, and run further validation.

**Comment by efeyazgan** [comment]
> I think we don't agree on the point that this should be part of validation.

Yes. Validation and request check. script have very different time-scales and purposes. 

> I don't want to have extra step in McM for GEN contact. It's too complicated even now (from the survey)

Now, independent of MCM we are asking them to run the script before they start validation anyway and then copy the output by hand to the google doc. 

> 
> As I said, if GEN contact want to waste their time, no local test should run. The test in McM should be just the final x-check. I don't see the point that if it passes the local, why it will fail in condor.
> 

I think we should discuss request checking script not the validation here. That's a separate issue. 


> I ran the script before I type the comment;
> (132) genproductions/bin/utils: python request_fragment_check.py --prepid HIG-RunIIFall17wmLHEGS-03157 --bypass_status
> ---> 1 requests will be checked:
> 
> if that fails, it because a release had been setup first ...
> Error: Cannot authenticate to: https://cms-pdmv-dev.cern.ch/mcm/
> 
> Error: Cannot authenticate to: https://cms-pdmv-int.cern.ch/mcm/
> 
> Error: Cannot authenticate to: https://cms-pdmv-dev.cern.ch/mcm/
> 
> Found a cookie file at /afs/cern.ch/user/s/srimanob/private/prod-cookie.txt. Make sure it's not expired!
> Using sso-cookie file /afs/cern.ch/user/s/srimanob/private/prod-cookie.txt
> HIG-RunIIFall17wmLHEGS-03157
> 
> HIG-RunIIFall17wmLHEGS-03157 Status= done
> GluGluToHHTo2B2WToLNu2J_node_SM_TuneCP5_13TeV-madgraph_correctedcfg
> /cvmfs/cms.cern.ch/phys_generator/gridpacks/2017/13TeV/madgraph/V5_2.4.2/GF_HH_SM/v1/GF_HH_SM_slc6_amd64_gcc481_CMSSW_7_1_30_tarball.tar.xz
> /eos/cms/store/group/phys_generator/cvmfs/gridpacks/2017/13TeV/madgraph/V5_2.4.2/GF_HH_SM/v1/GF_HH_SM_slc6_amd64_gcc481_CMSSW_7_1_30_tarball.tar.xz
> generate p p > h h
> 
> set group_subprocesses Auto
> set ignore_six_quark_processes False
> 
> [IMPORTANT WARNING] bwcutoff set to 15000.0. Note that large bwcutoff values can cause problems in production.
> 
>     * [WARNING] At least one of the MG5_aMC@NLO tmpdir patches is missing.
> 
>     * ```
>             --> Please check that you use:
>       ```
> 
>     * ```
>                 >=  CMSSW_7_1_32_patch1 in 7_1_X or
>       ```
> 
>     * ```
>                 >= CMSSW_9_3_9_patch1 in 9_3_X or
>       ```
> 
>     * ```
>                 >= 10_1_3 in 10_1_X or
>       ```
> 
>     * ```
>                 >= CMSSW_10_2_0_pre2 in 10_2_X.
>       ```
> 
>     * ```
>             Your request uses CMSSW_9_3_9_patch1 :
>       ```
> 
>     * ```
>             If you are not using a proper CMSSW version, please switch to that or
>       ```
> 
>     * ```
>             re-create the gridpack using the updated genproductions area
>       ```
> 
> 
> -------------------------MG5_aMC LO/MLM Many Threads Patch Check --------------------------------------
> 
>     * [OK] MG5_aMC@NLO LO nthreads patch OK in EOS
> 
>     * [OK] MG5_aMC@NLO LO nthreads patch OK in CVMFS
>       -------------------------EOF MG5_aMC LO/MLM Many Threads Patch Check ----------------------------------
> 
>     * [OK] no known inconsistency in the fragment w.r.t. the name of the dataset madgraph
> 
>     * [OK] Tune configuration probably OK in the fragment
> 
>     * [WARNING] Filters in the fragment but filter efficiency = 1

**Comment by efeyazgan** [comment]
> "If we mix it with validation for a 2 minute check we may need to wait for up to several days depending on the process. And in that case, it would be better not to have it in mcm then."
> is not what I propose.
> 
> If this part of validation fail, the validation job should quit from condor immediately. No need to download the genfragment, and run further validation.

OK, that may be a possibility but request checking script is being updated very frequently andin the scheme I am thinking without updating mcm (since it will just run a python command) mcm won't need an update. I am not sure that would be the case if we mix it with validation (@justinasr ?).

**Comment by efeyazgan** [comment]
> For the log, I think in the meeting we are talking about EOS. It should not be an issue to link to McM (but it's just a link to EOS file, and maybe can't show). It you need it to be shown in McM, something difference must be done.

It should be easy to find a solution for this part.

**Comment by srimanob** [comment]
> > "If we mix it with validation for a 2 minute check we may need to wait for up to several days depending on the process. And in that case, it would be better not to have it in mcm then."
> > is not what I propose.
> > If this part of validation fail, the validation job should quit from condor immediately. No need to download the genfragment, and run further validation.
> 
> OK, that may be a possibility but request checking script is being updated very frequently andin the scheme I am thinking without updating mcm (since it will just run a python command) mcm won't need an update. I am not sure that would be the case if we mix it with validation (@justinasr ?).

I don't propose to change anything in McM. What I said is that 
+ test: checkout the release
+ (NEW) test: clone GEN script from your branch and run
+ (NEW) test: detemine from the log => quit if fail
+ test: clone gen fragment and start validation

And this is one point in my original message that we should think about:
* We should avoid cloning 280MB of genproductions branch if not need.

**Comment by efeyazgan** [comment]
> > > "If we mix it with validation for a 2 minute check we may need to wait for up to several days depending on the process. And in that case, it would be better not to have it in mcm then."
> > > is not what I propose.
> > > If this part of validation fail, the validation job should quit from condor immediately. No need to download the genfragment, and run further validation.
> > 
> > 
> > OK, that may be a possibility but request checking script is being updated very frequently andin the scheme I am thinking without updating mcm (since it will just run a python command) mcm won't need an update. I am not sure that would be the case if we mix it with validation (@justinasr ?).
> 
> I don't propose to change anything in McM. What I said is that
> 
>     * test: checkout the release
> 
>     * (NEW) test: clone GEN script from your branch and run
> 
>     * (NEW) test: detemine from the log => quit if fail
> 
>     * test: clone gen fragment and start validation
> 

I did not understand what you are proposing above. To run it locally as being done now?

> 
> And this is one point in my original message that we should think about:
> 
>     * We should avoid cloning 280MB of genproductions branch if not need (following the recipe in MccM gdoc).

Good point. This is easy, we can put it somewhere else. Shall I already do it?

**Comment by qliphy** [comment]
One simple comment:   **no need to clone whole genproductions.  Only two files are needed**:

https://github.com/cms-sw/genproductions/blob/master/bin/utils/request_fragment_check.py

https://github.com/cms-sw/genproductions/blob/master/Utilities/scripts/update_gridpacks_mg242_thread.py

with a small modification [here](https://github.com/cms-sw/genproductions/blob/master/bin/utils/request_fragment_check.py#L339)

**Comment by qliphy** [comment]
It will be a small piece of work to insert those two python files,  and add a button in McM.   

As already discussed above,  script checking is not Mcm validation.  MC contacts can do this check before lauch McM validation.

**Comment by srimanob** [comment]
Please please understand from user point of view, not only the code development. The complication will fall to PdmV tool, and no one will blame Physics/GEN. I stand with my point, stay with a single validation click from user point of view. The intervention behind that "click" is PPD/PdmV/GEN job.

**Comment by pgunnell** [comment]
Dear all, sorry to be late here. I would also be in favour of adding a button (as I tried to describe at the beginning of this issue) but I see Phat's point. Here my poor opinion:

But I see two possible points:
- ADDING SCRIPT AS PART OF VALIDATION: if we include the script as part of the validation, and if validation is stopped if request checking script gives error, are we able to catch all possible errors?
- NOT ADDING A SCRIPT AS PART OF VALIDATION: without adding a button, would it be possible to submit two jobs in batch by clicking "next step", one job for the request script (which saves the log in Eos) and one job for the validation itself? I see Efe's point. If the log of the request script comes only at the end of the full validation, then we will not gain much in speeding up approval and submission procedure.

I would be in favour of the second option, if possible.

**Comment by srimanob** [comment]
Why the log of validation comes only at the full validation? If it fails the test since the beginning, no need to run further. If the request fails in GEN script, is that still good to go further?

**Comment by efeyazgan** [comment]
Even if we make it part of validation, at this point we can not really have a single OK from the request checking script. It is very dynamic. We can have cases where it can definitely be not OK but we can not guarantee for all.

**Comment by pgunnell** [comment]
I agree! (was writing at the same time)

**Comment by efeyazgan** [comment]
Hi, here is what we agreed on with Phat:
![req_check_script_in_mcm](https://user-images.githubusercontent.com/4159412/53161264-f664a380-35c9-11e9-97f9-c30eab6ee04c.png)
In words: Request checking script is run before validation. If it gives OK or warning it proceeds to validation. If it gives error then it stops. If the script gives warning it still runs validation because half of the cases it may not be a problem and we should not have MC contacts lose for that. New developments in the checking script may be first added as warning to be able to test. 
This doesn't require an extra button but thanks to the warning state still running validation it is not disruptive. 

The log is written in any case and to gitlab. The advantage of gitlab (or github) may be to have the discussion about the request in a single place. 

Does everyone agree with this plan?

**Comment by pgunnell** [comment]
Fine for me! Thanks!

**Comment by srimanob** [comment]
Hi,
Thx Efe for summary. To be clear, this will happen under a single click of request "validation". No extra click is needed.

**Comment by qliphy** [comment]
@efeyazgan   Thanks the summary workflow looks fine to me.   For gitlab/github, do you mean a new area or same as https://github.com/cms-PdmV ?   

"The log is written in any case and to gitlab."   How will make this automatically?

**Comment by efeyazgan** [comment]
We can discuss this in two weeks. In the meanwhile, we should update the code to have only OK, warning, and error codes and the error codes. We can move some of the errors to warning and vice versa.

**Comment by justinasr** [comment]
NOTE TO SELF: change `git clone ...` to `wget https://raw.githubusercontent.com/cms-sw/genproductions/master/bin/utils/request_fragment_check.py` and always add this to get_test command.
